graphql_compiler/schema_transformation/split_query.py

```diff
@@ -39,6 +47,165 @@ def __init__(self, query_ast):
+        targeting a specific schema. The second element is the set of all intermediate output
+        names that are to be removed at the end
+
+    Raises:
+        - GraphQLValidationError if the query doesn't validate against the schema, contains
+          unsupported directives, some property field occurs after a vertex field in some
+          selection, or some inline fragment coexists with other fields or inline fragment
+          on the same scope
+        - SchemaStructureError if the input merged_schema_descriptor appears to be invalid
+          or inconsistent
+    """
+    check_query_is_valid_to_split(merged_schema_descriptor.schema, query_ast)
+
+    # If schema directives are correctly represented in the schema object, type_info is all
+    # that's needed to detect and address stitching fields. However, GraphQL currently ignores
+    # schema directives when converting an AST to a schema object. Until this issue is
+    # fixed, it's necessary to use additional information from pre-processing the schema AST
+    edge_to_stitch_fields = _get_edge_to_stitch_fields(merged_schema_descriptor)
+    name_assigner = IntermediateOutNameAssigner()
+
+    root_query_node = SubQueryNode(query_ast)
+    query_nodes_to_split = [root_query_node]
+
+    # Construct full tree of SubQueryNodes in a dfs pattern
+    while len(query_nodes_to_split) > 0:
+        current_node_to_split = query_nodes_to_split.pop()
+
+        _split_query_one_level(current_node_to_split, merged_schema_descriptor,
+                               edge_to_stitch_fields, name_assigner)
+
+        query_nodes_to_split.extend(
+            child_query_connection.sink_query_node
+            for child_query_connection in current_node_to_split.child_query_connections
+        )
+
+    return root_query_node, frozenset(name_assigner.intermediate_output_names)
+
+
+def _get_edge_to_stitch_fields(merged_schema_descriptor):
+    """Get a map from type/field of each cross schema edge, to the fields that the edge stitches.
+
+    This is necessary only because GraphQL currently doesn't process schema directives correctly.
+    Once schema directives are correctly added to GraphQLSchema objects, this part may be
+    removed as directives on a schema field can be directly accessed.
+
+    Args:
+        merged_schema_descriptor: MergedSchemaDescriptor namedtuple, containing a schema AST
+                                  and a map from names of types to their schema ids
+
+    Returns:
+        Dict[Tuple(str, str), Tuple(str, str)], mapping (type name, vertex field name) to
+        (source field name, sink field name) used in the @stitch directive, for each cross
+        schema edge
+    """
+    edge_to_stitch_fields = {}
+    for type_definition in merged_schema_descriptor.schema_ast.definitions:
+        if isinstance(type_definition, (
+            ObjectTypeDefinition, InterfaceTypeDefinition
```

Comment:
Would this work if the stitch directives are provided as part of type/interface extension definitions?

For example:
```graphql
extend type Foo {
    out_Foo_Bar: [Bar] @stitch(source_field: "id", sink_field: "uuid")
}
```

Reply:
The code in its current form wouldn't work with that. I can extend it if desired, but something to note is that `merge_schemas` does not allow type extensions in the input schemas, so we don't expect there to be type extensions in the merged schema AST.

Reply:
That's okay then, don't worry about it. We can change it when we support them end-to-end.